src/realm/object-store/sync/sync_session.cpp

```diff
@@ -579,12 +579,12 @@ void SyncSession::handle_fresh_realm_downloaded(DBRef db, Status status,
         lock.unlock();
         if (auto local_subs_store = get_flx_subscription_store()) {
             // In DiscardLocal mode, only the active subscription set is preserved
-            // this means that we have to remove all other subscriptions including later
+            // This means that we have to remove all other subscriptions including later
             // versioned ones.
             auto mut_sub = local_subs_store->get_active().make_mutable_copy();
+            // Mark the new subscription set Complete because there must always exist an active subscription set.
+            mut_sub.update_state(sync::SubscriptionSet::State::Complete);
```

Comment:
This appears to work fine for this corner case. But I'm a little concerned that we are papering over a real underlying problem here. The subscription is actually not complete, it is in an error state. Is there some invariant of the subscription store that has been violated by setting the active subscription to an error?

If I undo these changes and run the test you added, it looks like the actual problem is that https://github.com/realm/realm-core/blob/master/src/realm/sync/subscriptions.cpp#L727-L730 where an assumption has been made that there is an object with primary key 0. This is probably true in the initial state, but not later on. I wonder if it is possible to hit this in normal flow somehow?

Reply:
If I understand correctly, the assumption is that there is always going to be a subscription with pk==0. If that is true, we should not remove it during a superceed state change. What do you think of these changes instead? https://github.com/realm/realm-core/compare/dt/fix_client_reset_crash...js/flx-client-reset?expand=1

Reply:
So the assumption is that there is always going to be a subscription with pk==0 **if** there is no other complete subscription. And that makes sense when opening a fresh realm. But once we have a complete subscription, we supercede and hence remove all the previous ones. I had the same idea to keep around the subscription with the primary key 0, but I think it's wrong. We always supercede subscriptions when one is marked Complete (https://github.com/realm/realm-core/blob/master/src/realm/sync/subscriptions.cpp#L393-L398). If one is marked Error, we should still have the last complete one. And that I think that should also be the case here. The new subscription is just a copy of the active/complete one, so it should not be in Error state since there is no issue with it.

Reply:
> If I undo these changes and run the test you added, it looks like the actual problem is that https://github.com/realm/realm-core/blob/master/src/realm/sync/subscriptions.cpp#L727-L730 where an assumption has been made that there is an object with primary key 0. This is probably true in the initial state, but not later on. I wonder if it is possible to hit this in normal flow somehow?

I cannot think of any other scenario. `supercede_prior_to` is called only when a subscription is marked complete, while `supercede_all_except` is the issue here (called only in this specific case)

Reply:
Okay, what you say about removing the pk==0 subscription makes sense, ignore my suggested change for that.
I am just wary of setting the subscription to complete when that is not the actual state of things, I feel that this may create other edge case bugs. I've tried to come up with a scenario where that matters, but am having a hard time.

Now I am questioning if it is correct to be superceeding subscriptions in this way at all. It may be fine for discard local, but I think it may be wrong for recovery mode in a scenario where the server breaks the schema but then changes it back such that FLX queries start to work again. In that situation we wouldn't want to superceed subscriptions made offline because those should be recovered.

Reply:
I wondered why we don't leave the subscriptions as they are..

Reply:
Superceeding the subscriptions will lead to compensating writes. We could superceed them for DiscardLocal and keep them for the recovery mode. I would keep them in all modes (in the worst case we'll create subscriptions for tables or objects that don't exist, but that's not an issue for the server afaik)

Reply:
I think you are right, and we should just delete https://github.com/realm/realm-core/blob/master/src/realm/object-store/sync/sync_session.cpp#L580-L589 and nuke the method `supercede_all_except()`. The comment says that the intent is to remove all later versioned (made offline) subscriptions, but that isn't necessary in the case of a client reset at this point in discard mode and it is plain wrong in recovery mode.
I think you can test the recovery scenario with something like this:

1. Subscribe to a query _A_ and successfully sync with the server.
2. Go offline and make a subscription _B_ and add an object that falls under that subscription.
3. Break the server schema so that subscription _A_ becomes an invalid query and terminate/restart the server.
4. Client reset with Recovery. This should fail _but not remove FLX query B_
5. Break the server schema again such that subscription _A_ becomes valid again and terminate/restart the server.
6. Reconnect the client, it should do a client reset with recovery successfully and subscription _B_ is now active and the object added in step 2 is uploaded to the server.

Reply:
Thanks for the test suggestion. I'll give it a try.

Reply:
I looked at https://github.com/realm/realm-core/blob/master/test/object-store/sync/flx_sync.cpp#L754-L800 for inspiration, but the outcome is not always deterministic. So, I added a new test based on the other test I added for this PR to test for this specific case.